ztest/ztest.go

```diff
@@ -121,66 +169,114 @@ func Run(t *testing.T, dirname string) {
 			if err != nil {
 				t.Fatalf("%s: %s", filename, err)
 			}
-			out, errout, err := run(zq, zt.ZQL, zt.OutputFormat, zt.OutputFlags, zt.Input...)
-			if err != nil {
-				if zt.errRegex != nil {
-					if !zt.errRegex.Match([]byte(errout)) {
-						t.Fatalf("%s: error doesn't match expected error regex: %s %s", filename, zt.ErrorRE, errout)
-					}
-				} else {
-					if out != "" {
-						out = "\noutput:\n" + out
-					}
-					t.Fatalf("%s: %s%s", filename, err, out)
-				}
-			} else if zt.errRegex != nil {
-				t.Fatalf("%s: no error when expecting error regex: %s", filename, zt.ErrorRE)
-			}
-			expectedOut, oerr := zt.getOutput()
-			require.NoError(t, oerr)
-			if out != expectedOut {
-				a := expectedOut
-				b := out
+			run(t, testname, bindir, dirname, filename, zq, zt)
+		})
+	}
+}
 
-				if !utf8.ValidString(a) {
-					a = encodeHex(a)
-					b = encodeHex(b)
-				}
+type File struct {
+	// Name is the name of the file in which the test shell code runs.
+	Name string `yaml:"name"`
+	// Data, Hex, and Source represents the different ways file data can
+	// be defined for this file.  Data is a string turned into the contents
+	// of the file, Hex is hex decded, and Source is a string representing
+	// the pathname of a file the repo that is read to comprise the data.
+	Data   string `yaml:"data,omitempty"`
+	Hex    string `yaml:"hex,omitempty"`
+	Source string `yaml:"source,omitempty"`
+	// Re is a regular expression describing the contents of the file,
+	// which is only applicable to output files.
+	Re string `yaml:"regexp,omitempty"`
+}
 
-				diff, _ := difflib.GetUnifiedDiffString(difflib.UnifiedDiff{
-					A:        difflib.SplitLines(a),
-					FromFile: "expected",
-					B:        difflib.SplitLines(b),
-					ToFile:   "actual",
-					Context:  5,
-				})
-				t.Fatalf("%s: expected and actual outputs differ:\n%s", filename, diff)
-			}
-			if err == nil && errout != zt.Warnings {
-				diff, _ := difflib.GetUnifiedDiffString(difflib.UnifiedDiff{
-					A:        difflib.SplitLines(zt.Warnings),
-					FromFile: "expected",
-					B:        difflib.SplitLines(errout),
-					ToFile:   "actual",
-					Context:  5,
-				})
-				t.Fatalf("%s: expected and actual warnings differ:\n%s", filename, diff)
-			}
-		})
+func (f *File) check() error {
+	cnt := 0
+	if f.Data != "" {
+		cnt++
 	}
+	if f.Hex != "" {
+		cnt++
+	}
+	if f.Source != "" {
+		cnt++
+	}
+	if cnt > 1 {
+		return fmt.Errorf("%s: must at most one of data, hex, or source", f.Name)
+	}
+	return nil
+}
+
+type Regexp struct {
```

Comment:
The source text for the regex is available via Regexp.String():
https://golang.org/pkg/regexp/#Regexp.String
so the wrapping struct here isn't necessary.

Reply:
sweet